File: LeetCode/Hard/3579-minimum-steps-to-convert-string-with-operations/3579-minimum-steps-to-convert-string-with-operations.py
class Solution:
    def minOperations(self, word1: str, word2: str) -> int:
        def count_swaps_and_replaces(source: str, target: str) -> int:
            pair_mismatches = [[0] * 26 for _ in range(26)]
            for s, t in zip(source, target):
                if s != t:
                    pair_mismatches[ord(s) - ord('a')][ord(t) - ord('a')] += 1
            
            operations_count = 0
            for i in range(26):
                for j in range(i + 1, 26):
                    operations_count += max(pair_mismatches[i][j], pair_mismatches[j][i])

            return operations_count

        dp = []
        for current_len in range(len(word1) + 1):
            min_operations = current_len
            for start in range(0, current_len):
                source = word1[start:current_len]
                target = word2[start:current_len]

                current_operations = dp[start] + min(count_swaps_and_replaces(source, target), 1 + count_swaps_and_replaces(source[::-1], target))
                min_operations = min(min_operations, current_operations)
                
            dp.append(min_operations)

        return dp[-1]

File: LeetCode/Hard/3579-minimum-steps-to-convert-string-with-operations/3579-minimum-steps-to-convert-string-with-operations.cpp
class Solution {
public:
    int minOperations(string word1, string word2) {
        vector dp(word1.length() + 1, 0);

        for (auto prefixLength = 1uz; prefixLength <= word1.length(); ++prefixLength)
        {
            auto minOperationsCount = prefixLength;
            for (auto start = 0uz; start < prefixLength; ++start)
            {
                string_view source(word1.data() + start, prefixLength - start);
                string_view target(word2.data() + start, prefixLength - start);

                auto nonReversedOperations = countReplacesAndSwaps(source, target);
                auto reversedOperations = 1 + countReplacesAndSwaps(source | views::reverse, target);

                minOperationsCount = min(dp[start] + min(nonReversedOperations, reversedOperations), minOperationsCount);
            }
            dp[prefixLength] = minOperationsCount;
        }

        return dp.back();
    }
private:
    template <typename R1, typename R2>
    size_t countReplacesAndSwaps(R1&& source, R2&& target)
    {
        array<array<int, 26>, 26> pair_mismatches = {};

        for (auto [s, t] : views::zip(source, target))
        {
            if (s != t)
            {
                pair_mismatches[s - 'a'][t - 'a']++;
            }
        }

        auto operationsCount = 0uz;
        for (auto s = 0uz; s < pair_mismatches.size(); ++s)
        {
            for (auto t = s + 1; t < pair_mismatches.size(); ++t)
            {
                operationsCount += max(pair_mismatches[s][t], pair_mismatches[t][s]);
            }
        }

        return operationsCount;
    } 
};